libedit introduced linker errors on aarch64 crossbuild

Symbols that are broken: `strlcpy`, `vis`, `tputs`, `tgoto`, `tgetent`, `tgetflag`, `tgetflag`, `tgetnum`, `tgetstr`, `strvis`, and `strunvis`.

Its clear to me the issue is caused by not being able to find libbsd. So I added the following to `.github/workflows/cmake-aarch64.yml `:

```
&& export LIBBSD_URL=http://ports.ubuntu.com/ubuntu-ports/pool/main/libb/libbsd/ \
        && curl $LIBBSD_URL 2>&1 | grep -oh "\"libbsd-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBBSD_URL/% \
        && dpkg-deb -R libbsd-dev*arm64.deb . \
        && rm -rf usr/share libbsd-dev*arm64.deb DEBIAN \
```

But it still isn't getting picked up. may have to explicitly pass in  `-lbsd` even though we don't have to do it on linux x86_64 hostOS? Before I do that will work to setup a local docker enviornment so testing this will be easier. 

Error trace:
```
FAILED: src/cli/Warf 
: && /usr/bin/clang++ --target=aarch64-unknown-linux-gnu --sysroot=/home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu -O3 -DNDEBUG -fuse-ld=lld src/cli/CMakeFiles/Warf.dir/main.cpp.o -o src/cli/Warf  src/lib/libWarfCore.a  -ledit && :
ld.lld: error: undefined symbol: strlcpy
>>> referenced by readline.o:(history_tokenize) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by readline.o:(get_history_event) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by readline.o:(history_expand) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced 9 more times
>>> did you mean: strncpy
>>> defined in: /usr/aarch64-linux-gnu/lib/libc.so.6

ld.lld: error: undefined symbol: vis
>>> referenced by readline.o:(rl_add_defun) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a

ld.lld: error: undefined symbol: tputs
>>> referenced by terminal.o:(terminal_move_to_line) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_move_to_line) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_move_to_line) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced 20 more times

ld.lld: error: undefined symbol: tgoto
>>> referenced by terminal.o:(terminal_move_to_line) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_deletechars) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_move_to_char) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced 5 more times

ld.lld: error: undefined symbol: tgetent
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch[64](https://github.com/farzonl/warfLang/actions/runs/3065624036/jobs/4949909214#step:7:65)-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a

ld.lld: error: undefined symbol: tgetflag
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced 3 more times

ld.lld: error: undefined symbol: tgetnum
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a

ld.lld: error: undefined symbol: tgetstr
>>> referenced by terminal.o:(terminal_set) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by terminal.o:(terminal_echotc) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a

ld.lld: error: undefined symbol: strvis
>>> referenced by hist.o:(hist_command) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by history.o:(history_save_fp) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by historyn.o:(history_save_fp) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a

ld.lld: error: undefined symbol: strunvis
>>> referenced by history.o:(history_w) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
>>> referenced by historyn.o:(history) in archive /home/runner/work/warfLang/warfLang/sysroots/aarch64-linux-gnu/usr/lib/aarch64-linux-gnu/libedit.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[33/34] Building CXX object test/CMakeFiles/WarfLang_TEST.dir/test.cpp.o
ninja: build stopped: subcommand failed.
```

--- a/.github/workflows/cmake-aarch64.yml
+++ b/.github/workflows/cmake-aarch64.yml
@@ -38,10 +38,6 @@ jobs:
         ca-certificates crossbuild-essential-arm64 \
         libstdc++-$GCC_VERS-dev-arm64-cross qemu-user wget rsync curl
     - name: setup sysroot
-      # Note I wanted to do  
-      # && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libtinfo-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
-      # However it looks like devs of libtinfo switched to zstd compression standard
-      # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=892664
       run: |
         cd ${{github.workspace}} \
         && mkdir -p sysroots/aarch64-linux-gnu/usr \
@@ -64,9 +60,9 @@ jobs:
         && dpkg-deb -R libbsd-dev*arm64.deb . \
         && rm -rf usr/share libbsd-dev*arm64.deb DEBIAN \
         && export LIBTINFO_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/n/ncurses/" \
-        && wget $LIBTINFO_URL/libtinfo-dev_6.2-0ubuntu2_arm64.deb \
-        && dpkg-deb -R libtinfo-dev*arm64.deb . \
-        && rm -rf usr/share libtinfo-dev*arm64.deb DEBIAN \
+        && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libncurses-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
+        && dpkg-deb -R libncurses-dev*arm64.deb . \
+        && rm -rf usr/share libncurses-dev*arm64.deb DEBIAN \
         && rsync -av usr ../ \
         && cd .. \
         && rm -rf tmp \

--- a/.github/workflows/cmake-aarch64.yml
+++ b/.github/workflows/cmake-aarch64.yml
@@ -59,6 +59,10 @@ jobs:
         && curl $LIBBSD_URL 2>&1 | grep -oh "\"libbsd-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBBSD_URL/% \
         && dpkg-deb -R libbsd-dev*arm64.deb . \
         && rm -rf usr/share libbsd-dev*arm64.deb DEBIAN \
+        && export LIBTINFO_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/n/ncurses/" \
+        && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libtinfo-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
+        && dpkg-deb -R libtinfo-dev*arm64.deb . \
+        && rm -rf usr/share libtinfo-dev*arm64.deb DEBIAN \
         && rsync -av usr ../ \
         && cd .. \
         && rm -rf tmp \

--- a/.github/workflows/cmake-aarch64.yml
+++ b/.github/workflows/cmake-aarch64.yml
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - master
+      - bugfix/issue/21
     paths:
       - '**.cpp'
       - '**.h'
@@ -81,8 +82,8 @@ jobs:
           -DCMAKE_ASM_COMPILER_TARGET=$TARGET_TRIPLE \
           -DCMAKE_EXE_LINKER_FLAGS_INIT="-fuse-ld=lld" \
           -DCMAKE_MODULE_LINKER_FLAGS_INIT="-fuse-ld=lld" \
-          -DCMAKE_SHARED_LINKER_FLAGS_INIT="-fuse-ld=lld" \
-          -DDISABLE_LIBEDIT=true
+          -DCMAKE_SHARED_LINKER_FLAGS_INIT="-fuse-ld=lld"
+
     - name: Build
       working-directory: ${{github.workspace}}/build
       shell: bash

--- a/.github/workflows/cmake-aarch64.yml
+++ b/.github/workflows/cmake-aarch64.yml
@@ -4,7 +4,6 @@ on:
   push:
     branches:
       - master
-      - bugfix/issue/21
     paths:
       - '**.cpp'
       - '**.h'

--- a/.github/workflows/cmake-aarch64.yml
+++ b/.github/workflows/cmake-aarch64.yml
@@ -38,6 +38,10 @@ jobs:
         ca-certificates crossbuild-essential-arm64 \
         libstdc++-$GCC_VERS-dev-arm64-cross qemu-user wget rsync curl
     - name: setup sysroot
+      # Note I wanted to do  
+      # && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libtinfo-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
+      # However it looks like devs of libtinfo switched to zstd compression standard
+      # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=892664
       run: |
         cd ${{github.workspace}} \
         && mkdir -p sysroots/aarch64-linux-gnu/usr \
@@ -60,7 +64,7 @@ jobs:
         && dpkg-deb -R libbsd-dev*arm64.deb . \
         && rm -rf usr/share libbsd-dev*arm64.deb DEBIAN \
         && export LIBTINFO_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/n/ncurses/" \
-        && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libtinfo-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
+        && wget $LIBTINFO_URL/libtinfo-dev_6.2-0ubuntu2_arm64.deb \
         && dpkg-deb -R libtinfo-dev*arm64.deb . \
         && rm -rf usr/share libtinfo-dev*arm64.deb DEBIAN \
         && rsync -av usr ../ \

--- a/Dockerfile
+++ b/Dockerfile
@@ -9,7 +9,12 @@ RUN apt-get update \
 # Set the working directory to root (ie $HOME)
 WORKDIR root
 
-COPY . .
+COPY test test
+COPY fuzz fuzz
+COPY CMakeLists.txt CMakeLists.txt
+COPY scripts scripts
+COPY cli-test-cases cli-test-cases
+COPY src src
 
 RUN cmake -B build -DCMAKE_BUILD_TYPE=Release \
     -DCMAKE_CXX_FLAGS="-coverage"  -DCMAKE_C_COMPILER=clang \

--- a/Dockerfile.aarch64
+++ b/Dockerfile.aarch64
@@ -1,0 +1,76 @@
+FROM ubuntu:22.04
+
+ENV DEBIAN_FRONTEND=noninteractive
+ENV GCC_VERS=12
+
+# Set the working directory to root (ie $HOME)
+WORKDIR root
+
+RUN apt-get update \
+    && apt install -y --no-install-recommends \
+               cmake ninja-build llvm-dev clang lld libc++-dev \
+               libc++abi-dev zlib1g-dev git build-essential \
+               ca-certificates crossbuild-essential-arm64 \
+               libstdc++-$GCC_VERS-dev-arm64-cross qemu-user \
+               wget rsync curl
+
+RUN cd \
+    && mkdir -p sysroots/aarch64-linux-gnu/usr \
+    && cd sysroots/aarch64-linux-gnu/usr \
+    && cp -r -v -L /usr/aarch64-linux-gnu/include /usr/aarch64-linux-gnu/lib . \
+    && cd lib \
+    && cp -r -v -L /usr/lib/gcc-cross/aarch64-linux-gnu/$GCC_VERS/*gcc* . \
+    && cp -r -v -L /usr/lib/gcc-cross/aarch64-linux-gnu/$GCC_VERS/*crt* . \
+    && cp -r -v -L /usr/lib/gcc-cross/aarch64-linux-gnu/$GCC_VERS/libsupc++.a . \
+    && cp -r -v -L /usr/lib/gcc-cross/aarch64-linux-gnu/$GCC_VERS/libstdc++*  . \
+    && cd ../../ \
+    && mkdir tmp \
+    && cd tmp \
+    && export LIBEDIT_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/libe/libedit/" \
+    && curl $LIBEDIT_URL 2>&1 | grep -oh "\"libedit-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBEDIT_URL% \
+    && dpkg-deb -R libedit-dev*arm64.deb . \
+    && rm -rf usr/share libedit-dev*arm64.deb DEBIAN \
+    && export LIBBSD_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/libb/libbsd/" \
+    && curl $LIBBSD_URL 2>&1 | grep -oh "\"libbsd-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBBSD_URL% \
+    && dpkg-deb -R libbsd-dev*arm64.deb . \
+    && rm -rf usr/share libbsd-dev*arm64.deb DEBIAN \
+    && export LIBTINFO_URL="http://ports.ubuntu.com/ubuntu-ports/pool/main/n/ncurses/" \
+    && curl $LIBTINFO_URL 2>&1 | grep -oh "\"libncurses-dev.*arm64.deb\"" | sort -nr | head -n1 | xargs -I% wget $LIBTINFO_URL/% \
+    && dpkg-deb -R libncurses-dev*arm64.deb . \
+    && rm -rf usr/share libncurses-dev*arm64.deb DEBIAN \
+    && rsync -av usr ../ \
+    && cd .. \
+    && rm -rf tmp \
+
+    # Clean up
+    && apt-get autoremove -y \
+    && apt-get clean -y \
+    && rm -rf /var/lib/apt/lists/*
+
+
+COPY src src
+COPY test test
+COPY fuzz fuzz
+COPY CMakeLists.txt CMakeLists.txt
+COPY scripts scripts
+COPY cli-test-cases cli-test-cases
+
+ENV TARGET_TRIPLE=aarch64-unknown-linux-gnu
+
+RUN cmake -S . -B build  \
+          -GNinja -DCMAKE_BUILD_TYPE=Release \
+          -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
+          -DCMAKE_ASM_COMPILER=clang \
+          -DCMAKE_INSTALL_PREFIX=/ \
+          -DCMAKE_SYSROOT=$HOME/sysroots/aarch64-linux-gnu \
+          -DCMAKE_C_COMPILER_TARGET=$TARGET_TRIPLE \
+          -DCMAKE_CXX_COMPILER_TARGET=$TARGET_TRIPLE \
+          -DCMAKE_ASM_COMPILER_TARGET=$TARGET_TRIPLE \
+          -DCMAKE_EXE_LINKER_FLAGS_INIT="-fuse-ld=lld" \
+          -DCMAKE_MODULE_LINKER_FLAGS_INIT="-fuse-ld=lld" \
+          -DCMAKE_SHARED_LINKER_FLAGS_INIT="-fuse-ld=lld"
+
+
+RUN ninja -C./build
+
+RUN qemu-aarch64 -L sysroots/aarch64-linux-gnu/usr/ build/test/WarfLang_TEST 

--- a/Dockerfile.aarch64
+++ b/Dockerfile.aarch64
@@ -47,13 +47,12 @@ RUN cd \
     && apt-get clean -y \
     && rm -rf /var/lib/apt/lists/*
 
-
-COPY src src
 COPY test test
 COPY fuzz fuzz
 COPY CMakeLists.txt CMakeLists.txt
 COPY scripts scripts
 COPY cli-test-cases cli-test-cases
+COPY src src
 
 ENV TARGET_TRIPLE=aarch64-unknown-linux-gnu
 

--- a/README.md
+++ b/README.md
@@ -63,3 +63,7 @@ run-clang-tidy.py -p build/ -header-filter='.*' -fix -format
 - Run: `docker run  --name test_vm -it warflang:latest`
 - Print out container id: `inspect test_vm -f '{{.Id}}'`
 - Get valgrind log: `docker cp <container id>:root/valgrind.log .`
+
+## Docker Build & Run aarch64
+- Build: `docker build -f Dockerfile.aarch64 -t warflang_arch64:latest .`
+- Run: `docker run  --name arm_test_vm -it warflang_arch64:latest`

--- a/src/cli/CMakeLists.txt
+++ b/src/cli/CMakeLists.txt
@@ -28,6 +28,9 @@ if(UNIX AND NOT isEmscripten AND NOT DISABLE_LIBEDIT)
     list(APPEND TargetLibraries
         "-lbsd"
     )
+    list(APPEND TargetLibraries
+        "-ltinfo"
+    )
 endif()
 
 target_link_libraries (${PROJECT_NAME}

--- a/src/cli/CMakeLists.txt
+++ b/src/cli/CMakeLists.txt
@@ -22,8 +22,11 @@ add_definitions(-DDISABLE_LIBEDIT=1)
 endif()
 
 if(UNIX AND NOT isEmscripten AND NOT DISABLE_LIBEDIT)
-    list(APPEND TargetLibraries 
+    list(APPEND TargetLibraries
         "-ledit"
+    )
+    list(APPEND TargetLibraries
+        "-lbsd"
     )
 endif()
 

--- a/src/cli/CMakeLists.txt
+++ b/src/cli/CMakeLists.txt
@@ -21,16 +21,23 @@ if(DISABLE_LIBEDIT)
 add_definitions(-DDISABLE_LIBEDIT=1)
 endif()
 
+if(UNIX AND NOT APPLE)
+    set(LINUX TRUE)
+endif()
+
 if(UNIX AND NOT isEmscripten AND NOT DISABLE_LIBEDIT)
     list(APPEND TargetLibraries
         "-ledit"
     )
-    list(APPEND TargetLibraries
-        "-lbsd"
-    )
-    list(APPEND TargetLibraries
-        "-ltinfo"
-    )
+    # Note we really only need this for the arm64 crossbuild
+    if(LINUX)
+        list(APPEND TargetLibraries
+            "-lbsd"
+        )
+        list(APPEND TargetLibraries
+            "-ltinfo"
+        )
+    endif()
 endif()
 
 target_link_libraries (${PROJECT_NAME}

